### PR TITLE
Wip/hughsie/client cert

### PR DESCRIPTION
**Pull request by hughsie**

### CODE REVIEW

**Comment by hughsie** [comment]
Urgh, s390 failed with:

    /fwupd/keyring{pkcs7}: (/build/build/src/fu-self-test:1759): FuKeyring-DEBUG: 17:18:49.843: generating a 3072 bit RSA private key...

Should I make that conditional on `g_test_slow()` or just disable it for s390? It's a shame, as the crypto code is exactly the kind of stuff I want checked for Endian-ness problems.

More generally, I wish there was a way to set up private build machines for Travis to use. I've got several chunky machines under my desk that sit there doing nothing 99% of the time.

**Comment by superm1** [comment]
> More generally, I wish there was a way to set up private build machines for Travis to use. I've got several chunky machines under my desk that sit there doing nothing 99% of the time.

I wonder if moving s390x to CircleCI might be better in this case?

File: src/fu-keyring-pkcs7.c [hunk 1]
```diff
@@ -22,60 +25,93 @@ struct _FuKeyringPkcs7
...
-		g_prefix_error (error, "failed to load %s: ", filename);
+	/* create certificate */
+	rc = gnutls_x509_crt_init (&crt);
+	if (rc < 0) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_SIGNATURE_INVALID,
+			     "crt_init: %s [%i]",
+			     gnutls_strerror (rc), rc);
 		return FALSE;
```
Comment (superm1) [inline src/fu-keyring-pkcs7.c]: should be NULL

File: src/fu-util.c [hunk 2]
```diff
@@ -2441,6 +2464,12 @@ main (int argc, char *argv[])
 		     /* TRANSLATORS: firmware approved by the admin */
 		     _("Sets the list of approved firmware."),
 		     fu_util_set_approved_firmware);
+	fu_util_cmd_array_add (cmd_array,
+		     "self-sign",
+		     "TEXT",
+		     /* TRANSLATORS: command description */
+		     _("Sign data using the client certificate"),
+		     fu_util_self_sign);
```
Comment (superm1) [inline src/fu-util.c]: Other than for testing the signing mechanism why put this in fwupdmgr?  I would think it makes more sense to have a `--signed` option and automatically sign relevant payloads it makes sense (or maybe even `--not-signed`) depending on the specific use case.

File: src/fu-keyring-pkcs7.c (hunk 1 above)
Comment (hughsie) [reply src/fu-keyring-pkcs7.c]: Good catch, thanks.

File: src/fu-util.c (hunk 2 above)
Comment (hughsie) [reply src/fu-util.c]: It was just for testing, but I agree it's a bit worthless if you can do the same kind of debugging in fwupdtool. I'll remove it, and add something like `--signed` for report uploading. If I get time today I'm going to look at the LVFS side to make sure all the signing stuff implemented as a client side cert works.

**Comment by superm1** [comment]
I think you also want to have fwupdagent signing the json payload by default (maybe invert the parameters though - add a `--no-sign` for that app)

**Comment by superm1** [comment]
Bash completion for `fwupdmgr` needs to recognize the new `--sign`

**Comment by superm1** [comment]
> Should I make that conditional on g_test_slow() or just disable it for s390? It's a shame, as the crypto code is exactly the kind of stuff I want checked for Endian-ness problems.

I think `g_test_slow` makes sense to add.  How about making it useful though:
* hide that test behind `g_test_slow`
* bump the timeout to something higher (60s instead of 30s) when run under `g_test_slow`?
* In CI mode (`CI=1` in environment)  launch the test suite with `-m slow`

**Comment by hughsie** [comment]
> I think you also want to have fwupdagent signing the json payload by default

Agreed, but that's a different PR :) I also think that `--no-sign` isn't required as it always needs to be secure.

> Bash completion for fwupdmgr needs to recognize the new --sign

Fixed, thanks.

> bump the timeout to something higher (60s instead of 30s)

Lets try doing this simple bit first; a 30s timeout is way too small when the build takes ~10 minutes anyway.

**Comment by superm1** [comment]
I don't really know how LGTM works for that CI failure.  It seems like it might be because of an older version of gnutls?  But I don't know for sure.

**Comment by hughsie** [comment]
So `gnutls_x509_spki_t` was added in gnutls 3.6.0, and according to https://packages.debian.org/search?keywords=libgnutls28-dev it means that stretch only has 3.5.8 :(

I guess three options would be:

 * Up the BR to 3.6.0 and disable the gnutls stuff on LGTM
 * Somehow try to build gnutls on CircleCI :/
 * Disable just the client key stuff on old GnuTLS versions

I'm erring on the last option. Agree?

**Comment by superm1** [comment]
So LGTM builds on stretch?  Is there any way to make it build on buster?

I think the last option makes most sense for flexibility of bringing fwupd to lots of places.  Bump up BR to 3.6.0 means it doesn't work in a lot of places anymore.

**Comment by superm1** [comment]
It seems odd to me - why isn't snap failing though?  It should be using the xenial gnutls which is even older ( 3.4.10-4ubuntu1.4)

**Comment by superm1** [comment]
Oh! `-Dpkcs7=false` on the snap build.  That explains it.

**Comment by hughsie** [comment]
> Bump up BR to 3.6.0 means it doesn't work in a lot of places anymore.

Agreed. I've also switched things around to only create the client certificate if we try to sign or verify something, as otherwise we'll fail to verify PKCS-7 metadata/firmware from an LVFS instance.

File: src/fu-keyring.c [hunk 3]
```diff
@@ -40,13 +40,33 @@ FuKeyringResult *
...
+{
+	FuKeyringClass *klass = FU_KEYRING_GET_CLASS (keyring);
+	g_return_val_if_fail (FU_IS_KEYRING (keyring), NULL);
+	g_return_val_if_fail (blob != NULL, NULL);
+	if (klass->sign_data == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "signing data is not supported");
+		return FALSE;
```
Comment (superm1) [inline src/fu-keyring.c]: Should be null

File: src/fu-tool.c [hunk 4]
```diff
@@ -1166,6 +1166,32 @@ fu_util_firmware_builder (FuUtilPrivate *priv, gchar **values, GError **error)
 	return fu_common_set_contents_bytes (values[1], firmware_blob, error);
 }
 
+static gboolean
+fu_util_self_sign (FuUtilPrivate *priv, gchar **values, GError **error)
```
Comment (superm1) [inline src/fu-tool.c]: Need the associated bash completion for this one now

Comment (hughsie) [reply src/fu-tool.c]: Fixed. I wonder if there's a way to fail to build/test/CI if we forget these? I know I'm really bad at remembering to update them...

Comment (superm1) [reply src/fu-tool.c]: Commands and flags are added (relatively) infrequently, I don't think it's likely worth the trouble of trying to come up with something like that.

**Comment by superm1** [comment]
LGTM

**Pull request merged**